include/boost/url/impl/url.ipp

```diff
@@ -584,23 +605,29 @@ url::
 set_userinfo(
     string_view s)
 {
-    detail::copied_strings buf(
-        this->string());
-    s = buf.maybe_copy(s);
-    check_invariants();
-    auto const n = pct_encode_bytes(
-        s, detail::userinfo_chars);
-    auto dest = set_userinfo_impl(n);
-    pct_encode(
-        dest,
-        get(id_host).data() - 1,
-        s,
-        detail::userinfo_chars);
-    decoded_[id_user] = s.size();
+    auto sep = s.find_first_of(':');
+    if (sep == string_view::npos)
+    {
+        set_user(s);
+        remove_password();
+    }
+    else
+    {
+        set_user(s.substr(0, sep));
+        set_password(s.substr(sep + 1));
```

Comment:
umm... this doesn't give the strong exception safety guarantee. You have to do the entire operation in a single resize. So you can't call the public-facing user functions. You need to call `resize_impl` (I think) and then write the encoded characters into that buffer. And then set `decoded_` correctly based on `sep`.

Reply:
Actually this is incorrect. `set_user` (line 611) will encode the colon and that's not what we want. Userinfo allows colon:
```
userinfo    = *( unreserved / pct-encoded / sub-delims / ":" )
```

so if they call `set_userinfo` with a colon, then a colon should appear in the string returned by `encoded_userinfo`. This function is actually very simple. Calculate the encoded size of `s` using the `userinfo_chars` set from above, call `resize_impl` on `[id_user, id_pass)` with that size (plus two for the double slashes if necessary), output the double slash (if needed) followed by percent-encoded chars, and then use `split` to cut the encoded string at the first colon.

Reply:
> You have to do the entire operation in a single resize

That makes sense.

I've updated the code to have the strong exception safety guarantee.

> This function is actually very simple

It was a little more than that but it worked

> set_user (line 611) will encode the colon

It won't. The caller never gives it a substr with the colon.

> a colon should appear in the string returned by encoded_userinfo

It will.

> then use split to cut the encoded string

_only_ `split` is problematic as it is because it breaks `decoded_`